src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    // NOTE:
+    /* A node table is the fundamental structure we use to track changes through the incremental 
+    generator api. It can be thought of as a series of slots that take their input from an
+    upstate table and produce 0-or-more outputs. When viewed from a downstream table the outputs
+    are presented as a single unified list, with each output forming the new input to the downstream
+    table.
+
+    Each slot has an associated state which is used to inform the operation that should be performed
+    to create or update the outputs. States generally flow through from upstream to downstream tables.
+    For instance an Added state implies that the upstream table produced a value that was not seen 
+    in the previous iteration, and the table should run whatever transform it tracks on the input 
+    to produce the outputs. These new outputs will also have a state of Added. A cached input specifies
+    that the input has not changed, and thus the outputs will be the same as the previous run. Added,
+    and Modified inputs will always run a transform to produce new outputs. Cached and Removed
+    entries will always use the previous entries and perform no work.
+
+    It is important to track Removed entries while updating the downstream tables, as an upstream 
+    remove can result in multiple downstream entries being removed. However, once all tables are up 
+    to date, the removed entries are no longer needed, and the remaining entries can be considered to
+    be cached. This process is called 'compaction' and results in the actual tables which are stored
+    between runs, as opposed to the 'live' tables that exist during an update.
+
+    Modified entries are similar to added inputs, but with a subtle difference. When an input is Added
+    all outputs are unconditionally added too. However when an input is modified, the outputs may still
+    be the same (for instance something changed elsewhere in a file that had no bearing on the produced
+    output). In this case, the state table checks the results against the previously produced values,
+    and any that are found to be the same instead get a cached state, meaning no new downstream work 
+    will be produced for them. Thus a modified input is the only slot that can have differing output 
+    states.
```

Comment:
This can be a bit hard to follow because it's fairly abstract. Can you try adding a concrete example at the bottom and walk through that? For example maybe talk about how a source file is processed by multiple generators (imagine two generators keyyng off diff attributes in the file) and how that would go through the table. 

Ideally though an example that would include a remove operation. That is the part I'm not quite following here.